File: python-sdk/harness_mem/langchain_memory.py
"""
NEXT-009: LangChain Memory 互換レイヤー（Python）

HarnessMemLangChainMemory は LangChain の BaseMemory と同等のインターフェースを実装し、
harness-mem をバックエンドに使った長期記憶を提供する。

LangChain BaseMemory 相当インターフェース:
- memory_variables: List[str]              — どの変数名を返すか
- save_context(inputs, outputs)           — 入力・出力をメモリに保存
- load_memory_variables(inputs) -> dict   — クエリに関連するメモリを取得
- clear()                                  — メモリをクリア（no-op）
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional, Sequence

from .client import HarnessMemClient


def _default_format_history(items: Sequence[Dict[str, Any]]) -> str:
    """デフォルトの履歴フォーマット: コンテンツを改行区切りで結合する。"""
    return "\n---\n".join(
        item.get("content", "") for item in items if item.get("content")
    )


@dataclass
class HarnessMemLangChainMemory:
    """
    LangChain の BaseMemory と互換性のある harness-mem メモリ実装。

    Example::

        from harness_mem import HarnessMemLangChainMemory

        memory = HarnessMemLangChainMemory(
            project="my-project",
            session_id="session-123",
        )
        # LangChain チェーンに渡す
        # chain = LLMChain(llm=llm, prompt=prompt, memory=memory)
    """

    project: str
    session_id: str
    base_url: str = "http://127.0.0.1:37888"
    token: Optional[str] = None
    timeout_sec: float = 8.0
    memory_key: str = "history"
    search_limit: int = 5
    format_history: Callable[[Sequence[Dict[str, Any]]], str] = field(
        default_factory=lambda: _default_format_history
    )

    def __post_init__(self) -> None:
        self._client = HarnessMemClient(
            base_url=self.base_url,
            timeout_sec=self.timeout_sec,
            token=self.token,
        )

    @property
    def memory_variables(self) -> List[str]:
        """LangChain BaseMemory.memory_variables: 返す変数名のリスト。"""
        return [self.memory_key]

    def save_context(
        self,
        inputs: Dict[str, Any],
        outputs: Dict[str, Any],
    ) -> None:
        """
        LangChain BaseMemory.save_context: 入出力ペアをメモリに保存する。
        input の内容と output の内容を結合してハーネスメムに記録する。
        """
        input_text = "\n".join(str(v) for v in inputs.values() if v)
        output_text = "\n".join(str(v) for v in outputs.values() if v)

        if not input_text and not output_text:
            return

        parts = []
        if input_text:
            parts.append(f"Human: {input_text}")
        if output_text:
            parts.append(f"AI: {output_text}")
        content = "\n".join(parts)

        self._client.record_checkpoint(
            session_id=self.session_id,
            title=input_text[:120] or "conversation turn",
            content=content,
            project=self.project,
        )

    def load_memory_variables(self, inputs: Dict[str, Any]) -> Dict[str, str]:
        """
        LangChain BaseMemory.load_memory_variables: クエリに関連するメモリを返す。
        inputs に含まれるテキストで harness-mem を検索し、関連する記憶を history 変数として返す。
        """
        query = " ".join(str(v) for v in inputs.values() if v).strip()

        if not query:
            return {self.memory_key: ""}

        try:
            response = self._client.search(
                query=query,
                project=self.project,
                limit=self.search_limit,
            )
            items = response.get("items", [])
            if not items:
                return {self.memory_key: ""}
            history = self.format_history(items)
            return {self.memory_key: history}
        except Exception:
            return {self.memory_key: ""}

    def clear(self) -> None:
        """
        LangChain BaseMemory.clear: メモリをクリアする。
        harness-mem はサーバー側で永続管理するため、クライアント側では no-op とする。
        """
        # no-op: harness-mem はサーバー側で管理される

File: sdk/src/langchain-memory.ts
/**
 * NEXT-009: LangChain Memory 互換レイヤー
 *
 * HarnessMemLangChainMemory は LangChain の BaseMemory と同等のインターフェースを実装し、
 * harness-mem をバックエンドに使った長期記憶を提供する。
 *
 * LangChain BaseMemory 相当インターフェース:
 * - memoryVariables: string[]      — どの変数名を返すか
 * - saveContext(inputs, outputs)   — 入力・出力をメモリに保存
 * - loadMemoryVariables(inputs)    — クエリに関連するメモリを取得
 * - clear()                        — メモリをクリア（harness-mem はサーバー管理のため no-op）
 */

import { HarnessMemClient } from "./client.js";
import type { HarnessMemClientOptions, SearchResultItem } from "./types.js";

export interface HarnessMemLangChainMemoryOptions extends HarnessMemClientOptions {
  /** ハーネスメムのプロジェクト名 */
  project: string;
  /** ハーネスメムのセッションID */
  session_id: string;
  /** 返却するメモリ変数名 (デフォルト: "history") */
  memoryKey?: string;
  /** 検索時に取得する最大件数 (デフォルト: 5) */
  searchLimit?: number;
  /** 検索結果をフォーマットする関数 */
  formatHistory?: (items: SearchResultItem[]) => string;
}

/**
 * LangChain の BaseMemory と互換性のある harness-mem メモリ実装。
 *
 * @example
 * ```typescript
 * const memory = new HarnessMemLangChainMemory({
 *   project: "my-project",
 *   session_id: "session-123",
 * });
 *
 * // LangChain チェーンに渡す
 * const chain = new LLMChain({ llm, prompt, memory });
 * ```
 */
export class HarnessMemLangChainMemory {
  private readonly client: HarnessMemClient;
  private readonly project: string;
  private readonly session_id: string;
  private readonly memoryKey: string;
  private readonly searchLimit: number;
  private readonly formatHistoryFn: (items: SearchResultItem[]) => string;

  constructor(options: HarnessMemLangChainMemoryOptions) {
    const { project, session_id, memoryKey, searchLimit, formatHistory, ...clientOptions } = options;
    this.client = new HarnessMemClient(clientOptions);
    this.project = project;
    this.session_id = session_id;
    this.memoryKey = memoryKey ?? "history";
    this.searchLimit = searchLimit ?? 5;
    this.formatHistoryFn = formatHistory ?? defaultFormatHistory;
  }

  /** LangChain BaseMemory.memoryVariables: 返す変数名のリスト */
  get memoryVariables(): string[] {
    return [this.memoryKey];
  }

  /**
   * LangChain BaseMemory.saveContext: 入出力ペアをメモリに保存する。
   * input の内容と output の内容を結合してハーネスメムに記録する。
   */
  async saveContext(
    inputs: Record<string, string>,
    outputs: Record<string, string>
  ): Promise<void> {
    const inputText = Object.values(inputs).filter(Boolean).join("\n");
    const outputText = Object.values(outputs).filter(Boolean).join("\n");

    if (!inputText && !outputText) {
      return;
    }

    const content = [
      inputText ? `Human: ${inputText}` : "",
      outputText ? `AI: ${outputText}` : "",
    ]
      .filter(Boolean)
      .join("\n");

    await this.client.record({
      project: this.project,
      session_id: this.session_id,
      event_type: "conversation_turn",
      payload: {
        title: inputText.slice(0, 120) || "conversation turn",
        content,
        observation_type: "context",
      },
    });
  }

  /**
   * LangChain BaseMemory.loadMemoryVariables: クエリに関連するメモリを返す。
   * inputs に含まれるテキストで harness-mem を検索し、関連する記憶を history 変数として返す。
   */
  async loadMemoryVariables(
    inputs: Record<string, string>
  ): Promise<Record<string, string>> {
    const query = Object.values(inputs).filter(Boolean).join(" ");

    if (!query.trim()) {
      return { [this.memoryKey]: "" };
    }

    const result = await this.client.search({
      query,
      project: this.project,
      session_id: this.session_id,
      limit: this.searchLimit,
    });

    if (!result.ok || result.items.length === 0) {
      return { [this.memoryKey]: "" };
    }

    const history = this.formatHistoryFn(result.items as SearchResultItem[]);
    return { [this.memoryKey]: history };
  }

  /**
   * LangChain BaseMemory.clear: メモリをクリアする。
   * harness-mem はサーバー側で永続管理するため、クライアント側では no-op とする。
   * 必要な場合は harness-mem の管理 API を直接使用してください。
   */
  async clear(): Promise<void> {
    // no-op: harness-mem はサーバー側で管理される
  }
}

/**
 * デフォルトの履歴フォーマット関数。
 * 検索結果を時系列順に連結したテキストとして返す。
 */
function defaultFormatHistory(items: SearchResultItem[]): string {
  return items
    .map((item) => item.content)
    .filter(Boolean)
    .join("\n---\n");
}
